resolver/resolver.cc

```diff
@@ -3970,6 +3970,94 @@ class ResolveSanityCheckWalk {
     }
 };
 
+class CheckAbstractClassInitialization {
+public:
+    struct CheckAbstractClassInitializationWalkResult {
+        vector<ast::ParsedFile> trees;
+    };
+
+    void postTransformSend(core::Context ctx, ast::ExpressionPtr &tree) {
+        auto &send = ast::cast_tree_nonnull<ast::Send>(tree);
+        if (send.fun != core::Names::new_()) {
+            return;
+        }
+
+        auto *id = ast::cast_tree<ast::ConstantLit>(send.recv);
+        if (id == nullptr || !id->symbol.exists() || !id->symbol.isClassOrModule()) {
+            return;
+        }
+
+        auto symbol = id->symbol.asClassOrModuleRef().data(ctx);
+        if (!symbol->flags.isAbstract) {
+            return;
+        }
+
+        auto singletonClass = symbol->lookupSingletonClass(ctx.state);
+        if (!singletonClass.exists()) {
+            return;
+        }
+
+        auto method_new = singletonClass.data(ctx)->findMethodTransitive(ctx.state, core::Names::new_());
+        // If the .new method we find is owned by Class, that means
+        // there was no user defined .new method, which warrants an error.
+        if (method_new.data(ctx)->owner == core::Symbols::Class()) {
+            if (auto e = ctx.beginError(send.loc, core::errors::Resolver::AbstractClassInstantiated)) {
+                e.setHeader("Attempt to instantiate abstract class `{}`", id->symbol.show(ctx));
+            }
+        }
```

Comment:
Thinking about this a little more, this is basically trying to implement method lookup and checks on the call, which is what `core/types/calls.cc` is doing.  Perhaps this check could be implemented there, instead, in `Class_new` or `Magic_selfNew`.

Reply:
`calls.cc` implementation can't cover the case outlined in https://github.com/sorbet/sorbet/pull/2222#issuecomment-558753052 so I decided to go with `validator.cc` approach. I'm not super convinced Sorbet shouldn't error in the case linked but it'd be a tough migration and would stray away from `sorbet-runtime` behaviour.

Offline we talked about the `resolver` approach as well however, that doesn't allow the error to be reported on the fast path. Also it resulted in a 80ms check in our monolith compared to a 3% increase with the `validator.cc` implementation. It also makes sense to me that we check for abstract on the same walk as `override` and `final` checks.